snapshots/wb/2023-08-14/pip_api.py

```diff
@@ -501,11 +501,14 @@ def generate_consolidated_percentiles(df):
     percentiles = range(1, 100, 1)
     df_percentiles = pd.DataFrame()
 
-    # Estimate percentiles using multiprocessing
-    with concurrent.futures.ProcessPoolExecutor() as executor:
-        futures = [executor.submit(calculate_percentile, p, df) for p in percentiles]
+    # Estimate percentiles using concurrency
+    futures = []
+    with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
```

Comment:
Here I applied the same principle, @Marigold, but the `calculate_percentile` function takes a lot. I tried replacing with `ProcessPoolExecutor` and the log went everywhere, to previous steps. How can this be even faster?

Reply:
For now I have set `max_workers` to 1, because it's too much RAM for the machine. That was probably the reason for this not working properly.

Reply:
Yeah, memory could be a problem. You can use smaller types in your dataframe. Either you do it automatically with
```
from owid.repack import repack_frame

def generate_consolidated_percentiles(df):
  ...
  df = repack_frame(df)
  ...
  # Estimate percentiles using concurrency
  ...
```
or you manually convert all columns `"ppp_version", "country", "year", "reporting_level", "welfare_type"` to categoricals with
```
df = df.astype({
  "ppp_version": "category",
  ...
})
```
Repacking / converting to categoricals should also speed up the groupby which probably takes the most time.

I can also try running the script myself and check for potential improvements. Should I run it on staging or download the files from staging and then run locally?

Reply:
Thanks @Marigold. Actually, waiting some hours for this is not a big deal, given the duration of the rest of the execution.

With my modifications to the code, I recommend you not to run it, because additional downloads from the API are necessary, so PIP wants to do that on their side. I will apply these changes and see if I need to change the code further afterward.